manifest/protobuf.go

```diff
@@ -137,23 +200,61 @@ func loadDescriptorSets(pkg *pbsubstreams.Package, manif *Manifest) ([]*desc.Fil
 			"https://buf.build",
 		)
 
-		request := connect.NewRequest(&reflectv1beta1.GetFileDescriptorSetRequest{
-			Module:  descriptor.Module,
-			Symbols: descriptor.Symbols,
-			Version: descriptor.Version,
-		})
+		var fileDescriptorSetResponse *descriptorpb.FileDescriptorSet
 
-		authToken := os.Getenv("BUFBUILD_AUTH_TOKEN")
-		if authToken != "" {
-			request.Header().Set("Authorization", "Bearer "+authToken)
-		}
+		// Don't use cache for mutable versions or empty versions
+		if !cache.shouldCache(descriptor.Version) {
+			request := connect.NewRequest(&reflectv1beta1.GetFileDescriptorSetRequest{
+				Module:  descriptor.Module,
+				Symbols: descriptor.Symbols,
+				Version: descriptor.Version,
+			})
 
-		fileDescriptorSet, err := client.GetFileDescriptorSet(context.Background(), request)
-		if err != nil {
-			return nil, fmt.Errorf("getting file descriptor set for %s: %w", descriptor.Module, err)
+			authToken := os.Getenv("BUFBUILD_AUTH_TOKEN")
+			if authToken != "" {
+				request.Header().Set("Authorization", "Bearer "+authToken)
+			}
+
+			fileDescriptorSet, err := client.GetFileDescriptorSet(context.Background(), request)
+			if err != nil {
+				return nil, fmt.Errorf("getting file descriptor set for %s: %w", descriptor.Module, err)
+			}
+
+			fileDescriptorSetResponse = fileDescriptorSet.Msg.FileDescriptorSet
+		} else {
+			// Try to load from cache first for valid semver versions
+			cacheKey := cache.generateCacheKey(descriptor.Module, descriptor.Version, descriptor.Symbols)
+			cachedFds, err := cache.load(cacheKey)
+
+			if err == nil {
+				// Cache hit
+				fileDescriptorSetResponse = cachedFds
+			} else {
+				// Cache miss, fetch from buf.build
+				request := connect.NewRequest(&reflectv1beta1.GetFileDescriptorSetRequest{
+					Module:  descriptor.Module,
+					Symbols: descriptor.Symbols,
+					Version: descriptor.Version,
+				})
+
+				authToken := os.Getenv("BUFBUILD_AUTH_TOKEN")
+				if authToken != "" {
+					request.Header().Set("Authorization", "Bearer "+authToken)
+				}
+
+				fileDescriptorSet, err := client.GetFileDescriptorSet(context.Background(), request)
```

Comment:
You should almost never use `context.Background`, in your `loadLocalProtobufs`, it should receive a `context` up until a caller has some context that make sense.

At the CLI level, we often use `cmd.Context()`